### PR TITLE
Modify package.json even with --skip-install

### PR DESCRIPTION
**Pull request by alfonso-presa**
- With this change running 'yo karma --skip-install & npm install' should
  work.
- Inspired by
  https://github.com/npm/npm/blob/master/lib/install.js#L408

Ment to solve this issue: https://github.com/yeoman/generator-karma/issues/74

### CODE REVIEW

File: app/index.js [hunk 1]
```diff
@@ -258,16 +258,47 @@ module.exports = yeoman.generators.Base.extend({
...
     }
+
+    this.options.plugins.push('grunt-karma');
+    if (this.options.coffee) {
+      this.options.plugins.push('coffee-script');
+    }
+
+    save(this.options.cwd || process.cwd(), this.options.plugins, this.async());
+    if (!this.options['skip-install']) {
+      this.on('end', function () {
```
Comment (SBoudrias) [inline app/index.js]: don't use `on('end')`

File: app/index.js [hunk 2]
```diff
@@ -258,16 +258,47 @@ module.exports = yeoman.generators.Base.extend({
...
+        this.npmInstall(this.options.plugins);
```
Comment (SBoudrias) [inline app/index.js]: If we edited package.json, there's no need to pass options to npm install.

File: app/index.js [hunk 3]
```diff
@@ -258,16 +258,47 @@ module.exports = yeoman.generators.Base.extend({
...
-        this.options.plugins.push('grunt-karma');
-        if (this.options.coffee) {
-          this.options.plugins.push('coffee-script');
+
+    function save (where, things, cb) {
+      var sortedObject = require("sorted-object");
+      var writeFileAtomic = require("write-file-atomic");
+      var saveTarget = path.resolve(where, "package.json");
+
+      fs.readFile(saveTarget, function (er, data) {
```
Comment (SBoudrias) [inline app/index.js]: This is not the way to handle files with Yeoman. See http://yeoman.io/authoring/file-system.html

File: app/index.js [hunk 4]
```diff
@@ -258,16 +258,47 @@ module.exports = yeoman.generators.Base.extend({
...
 
   installDeps: function () {
-    if (!this.options['skip-install']) {
-      this.on('end', function () {
-        this.options.plugins.push('grunt-karma');
-        if (this.options.coffee) {
-          this.options.plugins.push('coffee-script');
+
+    function save (where, things, cb) {
+      var sortedObject = require("sorted-object");
```
Comment (SBoudrias) [inline app/index.js]: move require to the top of the file.

File: app/index.js [hunk 5]
```diff
@@ -258,16 +258,47 @@ module.exports = yeoman.generators.Base.extend({
...
+
+        data["devDependencies"] = data["devDependencies"] || {};
+        Object.keys(things).forEach(function (t) {
+          data["devDependencies"][things[t]] = '*';
+        })
+
+        data["devDependencies"] = sortedObject(data["devDependencies"]);
+
+        data = JSON.stringify(data, null, 2) + "\n";
+        writeFileAtomic(saveTarget, data, function (er) {
```
Comment (SBoudrias) [inline app/index.js]: same as before, not the way to handle files in yeoman. See documentation reference.

File: app/index.js (hunk 1 above)
Comment (alfonso-presa) [inline app/index.js]: on('end') was already in the code. Anyway I will try to get rid of it ;-).

Comment (SBoudrias) [inline app/index.js]: That might be, don't take the code review personally - I swear I'm not judging your coding abilities on this. It's only about getting a better code and generator afterward.

We deprecated a lot of stuff and techniques over time, so I just find them as I do code review for our generators.

**Comment by alfonso-presa** [comment]
I've changed the PR to be more consistent in the way of writting files with the rest of the generator (I don't think it can use the this.fs without changing the whole generator, may be we should fill in an issue to update the code to the new API).

Also it is no longer using on('end'...

**Comment by SBoudrias** [comment]
You really cannot use `fs`. This will by pass our conflicter and differs from the expected behavior.

**Comment by alfonso-presa** [comment]
It's already being used in setupTravis, so all those side effects you say are already in the generator currently. If I understood correctly the API change implies an upgrade of the yeoman-generator dependency to at least 1.18.0 and also make a deep refactoring of the tests. I don't feel like this is something related to the issue I'm trying to solve. IMHO I think that the PR solves the issue and adds value to the project as is keeping the same quality standards than the current version. Upgrading the yeoman-generator version and changing to the new API should be a different task.

That said, I've been trying to make the upgrade to use 'this.fs'. It turns out to completely mess the tests. I tried to use inDir with callback to be able to generate the package.json file for the generator to add the dependencies and it breaks with a really weird error (may be is because I'm in a windows machine). May be you can give me a hand with this, should I fill an issue or am I doing something wrong.

When running this test:
```
  it('creates a travis file and updates package.json', function(done) {
    helpers
    .run(join(__dirname, '../app'))
    .withOptions({travis: true})
    .inDir(join(__dirname, 'temp2'), function(dir){
      require('fs').writeFileSync(join(dir,'package.json'), '{}');
    })
    .on('end', function () {
      assert.file(['.travis.yml']);
      assert.fileContent('package.json', /grunt test/);
      done();
    });
  });
```

I get this stack trace:

```
     Uncaught TypeError: Property 'undefined' of object spy is not a function
      at Conflicter.<anonymous> (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\lib\util\conflicter.js:187:36)
      at PromptUI.onCompletion (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\lib\ui\prompt.js:69:10)
      at AnonymousObserver.Rx.AnonymousObserver.AnonymousObserver.completed (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:1974:12)
      at AnonymousObserver.Rx.internals.AbstractObserver.AbstractObserver.onCompleted (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:1911:14)
      at AnonymousObserver.tryCatcher (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:568:29)
      at AutoDetachObserverPrototype.completed (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:4892:56)
      at AutoDetachObserver.Rx.internals.AbstractObserver.AbstractObserver.onCompleted (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:1911:14)
      at InnerObserver.onCompleted (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:3205:65)
      at InnerObserver.tryCatcher (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:568:29)
      at AutoDetachObserverPrototype.completed (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:4892:56)
      at AutoDetachObserver.Rx.internals.AbstractObserver.AbstractObserver.onCompleted (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:1911:14)
      at AutoDetachObserver.tryCatcher (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:568:29)
      at AutoDetachObserverPrototype.completed (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:4892:56)
      at AutoDetachObserver.Rx.internals.AbstractObserver.AbstractObserver.onCompleted (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:1911:14)
      at InnerObserver.onCompleted (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:3205:65)
      at InnerObserver.tryCatcher (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:568:29)
      at AutoDetachObserverPrototype.completed (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:4892:56)
      at AutoDetachObserver.Rx.internals.AbstractObserver.AbstractObserver.onCompleted (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:1911:14)
      at AutoDetachObserver.tryCatcher (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:568:29)
      at AutoDetachObserverPrototype.completed (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:4892:56)
      at AutoDetachObserver.Rx.internals.AbstractObserver.AbstractObserver.onCompleted (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\node_modules\rx\dist\rx.js:1911:14)
      at C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\lib\utils\utils.js:53:13
      at C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\lodash\index.js:7173:25
      at C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\node_modules\inquirer\lib\ui\prompt.js:98:7
      at Object._onImmediate (C:\Users\Alfonso\WorkSpace\forks\generator-karma\node_modules\yeoman-generator\lib\test\adapter.js:39:5)
      at processImmediate [as _immediateCallback] (timers.js:345:15)
```
any clues?

File: app/index.js [hunk 6]
```diff
@@ -187,6 +189,46 @@ module.exports = yeoman.generators.Base.extend({
...
+
+      try {
+        data = JSON.parse(content);
+      } catch (err) {
+        that.log.error('Could not parse package.json.', err);
+        done();
+        return;
+      }
+
+      data["devDependencies"] = data["devDependencies"] || {};
```
Comment (SBoudrias) [inline app/index.js]: `data.devDependencies`

**Comment by alfonso-presa** [comment]
You can check my WIP in the refactoring in this other branch:
https://github.com/alfonso-presa/generator-karma/tree/ChangeGeneratorVersion

**Comment by SBoudrias** [comment]
Hey, sorry about that. I really though all our generator at least used the legacy file system. I don't know anytime where fs was supposed to be used in generators.

Let's keep it this way for now.

**Comment by alfonso-presa** [comment]
It's ok :-). I completely understand your point.

File: app/index.js [hunk 7]
```diff
@@ -187,6 +189,46 @@ module.exports = yeoman.generators.Base.extend({
...
+      try {
+        data = JSON.parse(content);
+      } catch (err) {
+        that.log.error('Could not parse package.json.', err);
+        done();
+        return;
+      }
+
+      data.devDependencies = data.devDependencies || {};
+      Object.keys(that.options.plugins).forEach(function (t) {
```
Comment (SBoudrias) [inline app/index.js]: pretty plugins is an array. How about `this.options.plugins.forEach()`. Then you get the key directly and not an array index.

Comment (alfonso-presa) [inline app/index.js]: You're completely right sir. Fixed :-)

**Comment by SBoudrias** [comment]
Okay, squash those commits and we'll merge.

**Comment by alfonso-presa** [comment]
Ready to merge :-). By the way what about the problem with the inDir. I have the 0.19.1 almost done pending of that testing problem. Should I fill in an issue in yeoman generator or I'm I making something wrong?

**Comment by SBoudrias** [comment]
@alfonso-presa the conflicter try to write to the console. You should pass `force: true` in `withOptions`.

**Comment by alfonso-presa** [comment]
Thanks! That was it :-). I've already done a PR with the 0.19.1 upgrade.

**Pull request merged**